### PR TITLE
feat: support more node crypto functions

### LINKED ISSUE #91: Complete node specific `crypto` polyfills
(followup for #90)

### PR DESCRIPTION
**Pull request by AwesomeDude091**
Hopefully a fix for #110 
Resolves #91

### CODE REVIEW

**Comment by AwesomeDude091** [comment]
Are there any more changes required @pi0 ?

**Comment by pi0** [comment]
Thanks for the pull request dear @AwesomeDude091. This PR is great however we probably need to use more of webCrypto API as node:crypto is only available for Node.js environments and compatibility layer of cloudflare pages for node.

**Comment by Hebilicious** [comment]
@AwesomeDude091 @Romitou I would be surprised if this solves your issues with cloudflare *once it's deployed*.
Building locally will work because you are using a node.js environment for the build, however, what you're doing here isn't a polyfill, you're just using "node:crypto" in unenv (which should break in workerd as "node:crypto" won't be available).

If you want to polyfill all "node:crypto" method, you would basically need to do a real polyfill of node:crypto, similar to https://github.com/browserify/crypto-browserify 

What I can suggest instead is that instead of trying to polyfill `nuxt-auth` with cloudflare, you use another library that doesn't require node-only internals at runtime, such as https://github.com/Hebilicious/authjs-nuxt or Lucia https://lucia-auth.com/start-here/getting-started?nuxt or even Hanko https://github.com/nuxt-modules/hanko

**Review by Hebilicious — CHANGES_REQUESTED** [review]
You shouldn't import any node internals in the runtime. 
Type imports are fine.

**Comment by Hebilicious** [comment]
@pi0 we could go with an incremental approach for the crypto polyfills, starting with the functions needed for nuxt-auth seems less daunting.
We also need to make sure that we don't polyfill things that are available in crypto global

**Comment by AwesomeDude091** [comment]
Update nearly all Cloudflare Issues have been resolved with the current build (testing on a local wrangler). It builds successfully and without warnings, however the worker gets fails on initalization with error ` Uncaught TypeError: Cannot read properties of undefined (reading 'split')
  at do835gcosb7.js:20707:33`, however this seems to be more of a general issue with cloudflare and nuxt also outlined here:  https://github.com/cloudflare/workers-sdk/issues/3144
  
Any ideas on how to tackle that 'split' issue?  
  
P.S. For testing I build unenv locally paste it into my node_modules folder of my test application (overwritting the current released build) and then build and wrangler execute

**Comment by AwesomeDude091** [comment]
I isolated the line that is causing issues in the bundledWorker:
`var [Bv, Mv] = ge.versions.node.split(".").map((F2) => parseInt(F2, 10));`

**Comment by Hebilicious** [comment]
> I isolated the line that is causing issues in the bundledWorker: `var [Bv, Mv] = ge.versions.node.split(".").map((F2) => parseInt(F2, 10));`
> 
> ```
> var me = "undefined" != typeof globalThis ? globalThis : "undefined" != typeof self ? self : "undefined" != typeof global ? global : {};
> me.process = me.process || oe;
> var ge = me.process;
> ```
> 
> Do you guys @pi0 @Hebilicious have any ideas?

Read my comment above about using node internals in unenv. This approach won't work, you need to remove all imports from "node:crypto" and *polyfill* them. You can keep the type imports as they won't be included in the final bundle.

**Comment by AwesomeDude091** [comment]
I need some help, here it is not working out

**Comment by AwesomeDude091** [comment]
I cannot figure out the port for generateKey, generateKeySync, generateKeyPair, generateKeyPairSync

**Comment by AwesomeDude091** [comment]
> I cannot figure out the port for generateKey, generateKeySync, generateKeyPair, generateKeyPairSync

Any help with this would be much appreciated

**Comment by AwesomeDude091** [comment]
TODOs for a nearly complete Crypto Polyfill

- [ ] generateKey
- [ ] generateKeySync
- [ ] generateKeyPair
- [ ] generateKeyPairSync
- [ ] KeyObject

**Comment by AwesomeDude091** [comment]
@pi0 @Hebilicious Does the above list check out, are the previous new commits okay and is there any chance that you could give me a pointer as to what to do for the generateKey and generateKeyPair logic, especially the difference between sync and not sync?

**Review by Hebilicious — CHANGES_REQUESTED** [review]
Nice work @AwesomeDude091 !

You are attempting to do something that is extremely complicated. Supporting ALL of node crypto internals might be unachievable. 

Even if you succeed, this approach has some downsides : 

- Runtime size : Workers environments are designed to run with low resources. Even if you can implement the more complex crypto functionalities, they are CPU heavy and you might not be able to use them in a worker.

- Security : Poyfill these things could open your applications to vulnerability as you are now in charge of making it safe.

I'm not a fan of adding browserify as a dependency. For our use case, it might be better to copy their logic instead of using their package.

What I would like to make these extra polyfill opt-in, ie have them behind some kind of configuration flag such as `preset.extraCryptoPolyfill` ...

If you *still* wish to continue working on this, I can recommend you take a look at https://github.com/PeculiarVentures/webcrypto and their other repositories.

As I said before, I wouldn't do all of this *just* to use`sidebase/nuxt-auth` on cloudflare workers. I would only recommend you do this if you're really interested in cryptography and to find out how to do this, since you're in un-charted territory and it will be relatively hard to help you.

Bonus reading : 
https://tonyarcieri.com/whats-wrong-with-webcrypto

**Comment by pi0** [comment]
I will be working on this PR/Issue later this week once had time. 

In general overview for changes:

- Partially reimplementing Node API for web, is WHAT UNENV IS BULT FOR so this PR is not in a wrong direction 👍🏼 
- We shall not import directly from `node:` namespace (unless for types) and implement from `unenv/node/stream` for example for stream utils when we need
- When cloudflare and other runtimes have more stable polyfill, we shall use it (it is unrelated to this PR btw and more about higher order usage in Nitro)

File: src/runtime/node/util/types/_types.ts [hunk 1]
```diff
@@ -165,12 +166,10 @@ export const isBigInt64Array: typeof utilTypes.isBigInt64Array = notImplemented(
 export const isBigUint64Array: typeof utilTypes.isBigUint64Array =
   notImplemented("util.types.isBigUint64Array");
 
-// @ts-expect-error not implemented yet
-export const isKeyObject: typeof utilTypes.isKeyObject = notImplemented(
-  "util.types.isKeyObject"
-);
+// @ts-ignore
```
Comment (Hebilicious) [inline src/runtime/node/util/types/_types.ts:169]: should be ts-expect-error

File: src/runtime/node/crypto/index.ts [hunk 2]
```diff
@@ -1,8 +1,17 @@
 // https://nodejs.org/api/crypto.html
 // https://github.com/unjs/uncrypto
-import type nodeCrypto from "node:crypto";
+import type nodeCrypto from "node:crypto"
+import stream from "node:stream";
+// @ts-ignore
+import cryptoBrowserify from "crypto-browserify"
+// @ts-ignore
+import { webTimingSafeEqual } from "@advena/web-timing-safe-equal"
```
Comment (Hebilicious) [inline src/runtime/node/crypto/index.ts:8]: we should use the browserify version instead of this one https://www.npmjs.com/package/timing-safe-equal

**Review by Hebilicious — CHANGES_REQUESTED** [review]
- Some comments that needs to be removed/included.
- Should we proxy all of browserify-crypto and augment it with the correct types ?
- Should unenv have an explicit opt-in behaviour for these browserify crypto polyfills ? This would be consistent with something like https://github.com/cyco130/esbuild-plugin-polyfill-node
- What happens if only one import is needed ? Does *everything* gets added to the final bundle ? @pi0

**Review by Hebilicious — CHANGES_REQUESTED** [review]
duplicate

**Comment by pi0** [comment]
@Hebilicious Thanks for the review. I will handle the next steps on this. Kindly please NEVER request for change in other PRs in unjs repos unless you are an explicit maintainer of it ([read more about roles](https://github.com/unjs/governance/#unjs-team)). You can always ask questions about general project directions from maintainers.

> Should we proxy all of browserify-crypto and augment it with the correct types ?

Not necessarily for all. Only for ones are possible to be purely implemented in JS with Web APIs. For those that are not possible or not implemented yet, unenv simply throws an `Not Implemented` error. But this often helps to fix lots of build errors with incompatible Node.js libraries that simply import things and in practice don't essentially require them.

> Should unenv have an explicit opt-in behavior for these browserify crypto polyfills ? This would be consistent with something like https://github.com/cyco130/esbuild-plugin-polyfill-node

Unenv only provides presets and polyfill (today). It is up to the consumer (nitro cloudflare presets for instance) to apply opt-in/out behavior which has to be opt-out i guess since cloudflare (and only with a flag) allows some of Node apis.

> What happens if only one import is needed ? Does everything gets added to the final bundle ? 

Exports are tree-shakable in general unless a lib is using `require` with untreeshakable syntax or `import * from`.

**Comment by AwesomeDude091** [comment]
@pi0 Are there TransformOptions / StreamOptions implemenations as part of the UNENV polyfill?

**Comment by AwesomeDude091** [comment]
My bad, figured it out

**Comment by AwesomeDude091** [comment]
@pi0 I have started the implementation of  `generateKeyPair` hope this is the right direction...

**Comment by pi0** [comment]
Hi dear @AwesomeDude091 thanks again for working on all this and so sorry for all troubles. 

#121 adds all needed stubs so builds won't fail. 

For cloudflare (and other worker runtimes) we shall leverage [Node.js compatibility](https://developers.cloudflare.com/workers/runtime-apis/nodejs/)

If you think some of utils could be actually implemented and possible without external dependencies such as browserify-crypto and purely based on webcrypto/subtle, PR more than welcome on case by case basis.

Best luck ❤️

**Pull request closed without merging**